ships/0025-event-driven-builds.md

```diff
@@ -0,0 +1,300 @@
+      value: $(body.ref)
+```
+
+Similar `ClusterTriggerBinding`s need to be shipped for `Gitlab` and `BitBucket`.
+
+2. A "Custom Tekton Task" controller with the following behaviour:
+    * _watches_ `Tekton` `Run` resources referencing Shipwright's `Build` resources.
+    * Expects the commit ID and branch name in the `params`.
+    * Based on the above information, the controller would generate the following:
+          * A `Build` with the revision and output image tag overwritten with the above information based on the optional `.spec.webhook.imageTagPolicy`.
+          * A `BuildRun` referencing the above `Build`.
+
+
+
+#### Webhook endpoint creation process
+
+ 
+Upon creation of a `Build` resource by the user, the following would occur:
+
+1. The Shipwright Build Controller would create a `TriggerTemplate` that would map the information coming in from the event into the "parameters" 
+   expected by the `Run` resource ie, the branch name & the commit ID.
+   
+ ```
+ 
+### Generated by the `Build` reconciler. 
+ 
+apiVersion: triggers.tekton.dev/v1alpha1
+kind: TriggerTemplate
+metadata:
+ name: my-build-name
+spec:
+ params:
+   - name: branch
+   - name: commit
+ resourceTemplates:
+   - apiVersion: tekton.dev/v1alpha1
+     kind: Run
+     metadata:
+       generateName: build-execution-
+     spec:
+       ref:
+         apiVersion: shipwright.io/v1alpha1
+         kind: Build 
+         name: my-build-bame
+       timeout: 3000s
+       params:
+         - name: branch
+           value: $(tt.params.branch)
+         - name: commit
+           value: $(tt.params.commit)
+ ```
+
+2. The Shipwright Build Controller would create a `EventListener` under-the-hood per build. 
+    
+```
+### Generated by the `Build` reconciler. 
+
+apiVersion: triggers.tekton.dev/v1alpha1
+kind: EventListener
+metadata:
+ name: name-of-the-build
+spec:
+ serviceAccountName: pipeline
+ triggers:
+   - bindings:
+       - ref: github-shipwright-webhook
+     template:
+       name: name-of-the-build
+   
+```
+
+And that's it, you may now go ahead and 
+
+### Test Plan
+
+To be filled.
+
+### Release Criteria
+
+To be filled.
+
+#### Removing a deprecated feature [if necessary]
+
+N/A
+
+#### Upgrade Strategy [if necessary]
+
+N/A
+
+### Risks and Mitigations
+
+1. This feature opens up the possiblility of triggering Build executions for branches which weren't explicitly specified. 
+
+This isn't a risk per se since only repository committers would have the permissions needed to push branches. Therefore,
+as long as the image is tagged appropriately to indiciate that it is not built off 'main' ( or the branch specified iniially by the user in the `Build`), 
+this should not be a problem.
+
+While the enhancement could have been scoped to only deal with the branch/revision explicitly specified in the `Build` resource, 
+doing the same would have have excluded some key use cases where users push to a different branch before merging to 'main'. 
```

Comment:
Hm, but what is the user's expectation? I'd say if I specify a Build with a source revision set to main, that I only want commits into main to trigger a BuildRun.

Reply:
I did go in that direction and quickly found it to be too cumbersome to have to define a new `Build` for every branch. 🤷‍♂️

Reply:
I see. We'll have to think about a good solution. Let's look at the scenario first: in my experience, repositories typically only have one or two major branches (main and develop) - having two Builds for this would be okayish for me because of how we use those branches internally (the images go into different repos).

The other branches you have in mind, would not that have been pull requests ? Because that would make sense for me as a scenario: define branch A in the build and define that pull requests for that branch should also be built.

Another scenario where branches do not play any role would be to build a tag once it gets pushed.